### ISSUE TITLE
Create credible breaker for rollback scenario

### ISSUE THREAD

**Issue by mms710** [zcash/zcash#3854]
As part of the rollback protection and response work, we need to create a credible breaker for a rollback scenario

**Comment by leto** [comment]
What does this mean?

**Comment by zebambam** [comment]
Yeah sorry for the shorthand, I'll try to clarify the thinking around this based on a meeting recently. The idea is to have a system that monitors the network for rollbacks, and upon detecting a rollback beyond a certain depth, sends out an alert using the alert keys across the network that puts nodes that support this feature (which we hope will be all nodes) into safe mode so as not to execute further transactions. 

This is, in essence, deliberately converting the "failure mode" of the network in the case where some single entity controls enough hashpower to create a deep rollback in order to execute double (multiple) spends from arbitrary theft to denial of service.

It's a choice that we'd rather see the network temporarily stop being effective rather than have users (typically exchanges, but could be anyone executing transfers automatically) scammed out of their zcash.

This ticket shouldn't be considered alone, it should be considered within the context of the other tickets carrying the "rollback protection" label, and seen in stages, and also as a temporary measure against rollback attacks that we're seeing with other coins.

Before this change can be effective, we would first have to fix the known issues with safe mode and the alert system (ticket #3856), let everyone know (ticket #3855) that this system exists and that they should expect a network alert that puts their node into safe mode, and they should expect to recover from safe mode socially and manually (ticket #3853).

In terms of which chain to follow, if we select the depth well enough and the alert propagates fast enough across the network, it shouldn't matter which chain we pick, just that it results in no double spends and the user experience for making transfers is transparent and useful. Now that I come to explain this proposal, I think we should have some user experience designers weigh in on how to reflect these changes at the UX level.

**Comment by tromer** [comment]
Lost you at "it shouldn't matter which chain we pick". One chain would contain valid user transactions (which could be arbitrarily harmful to revert, especially if the depth is larger than common-practice "confirmation time"). The other chain may contain double-spend attack transactions, and maybe some _other_ valid user transactions. How does it not matter which chain becomes the new consensus?

**Comment by zebambam** [comment]
It does matter, but only in the sense that it needs to not contain / have reverted spends. I'm trying to describe a rule that will work even if an attacker creates multiple competing chains within the confirmation / rollback breaker depth in an attempt to confuse users. In that case, the breaker trips and social consensus only needs to select a winning chain that does not result in double spends.

**Comment by tromer** [comment]
Let me paraphrase to make sure we're on teh same page:

Any _consistent_ choice of _valid_ chain fork will _avoid double spends_. This is guaranteed by the protocol (valid chains can't contain the same note nullifier more than once). So to avoid double spends, it will suffice to have an out-of-band social mechanism for deciding on the consensus chain fork as part of recovery from the rollback alert.

But this won't absolve the "social consensus" from making painful judgment call about the _other_ merits of the candidate chain forks, like which one is more likely to have harmful-to-revert transactions (such as paying for by-now-delivered merchandise, or depositing by-now-withdrawn funds into an exchange).

**Comment by zebambam** [comment]
Right. The social consensus will have to be pre-built as much as possible, and the method for deciding it I think should be transparent, repeatable, deterministic for the most part. It needs to be communicated (ticket #3855) and it needs to be based on valid information. One way to do this is notaries, although ticket #3782 is at this point a bit overly complex, so the other piece of this puzzle is to have @daira (get under that bus!) detail a simple and effective notary system that will feed data into both the circuit breaker in its second iteration perhaps and the social consensus.

**Comment by zebambam** [comment]
"But this won't absolve the "social consensus" from making painful judgment call about the _other_ merits of the candidate chain forks, like which one is more likely to have harmful-to-revert transactions (such as paying for by-now-delivered merchandise, or depositing by-now-withdrawn funds into an exchange)."

So for this consideration, the depth of the breaker depth needs to be lower than the confirmation depth so that nobody should have made irreversible changes.
